pkg/kv/kvserver/kvserverbase/raftversion.go

```diff
@@ -13,34 +13,52 @@ package kvserverbase
+	// regular Raft traffic.
+	//
+	// This is a raftpb.Entry of type EntryNormal whose Data slice is either
+	// empty or whose first byte matches RaftVersionStandard. The subsequent
+	// eight bytes represent a CmdIDKey. The remaining bytes represent a
+	// kvserverpb.RaftCommand.
 	RaftVersionStandard RaftCommandEncodingVersion = 0
-	// RaftVersionSideloaded indicates a proposal containing an SSTable which
-	// preferably should be sideloaded (i.e. not stored in the Raft log
-	// wholesale). Can be treated as a regular proposal when arriving on the wire,
-	// but when retrieved from the local Raft log it necessary to inline the
-	// payload first as it has usually been sideloaded.
+	// RaftVersionSideloaded indicates a proposal representing the result of a
+	// roachpb.AddSSTableRequest for which the payload (the SST) is stored outside
+	// the LSM to improve storage performance.
+	//
+	// This is a raftpb.Entry of type EntryNormal whose data slice is either
+	// empty or whose first byte matches RaftVersionSideloaded. The remaining
+	// bytes represent a kvserverpb.RaftCommand whose kvserverpb.ReplicatedEvalResult
+	// holds a nontrival kvserverpb.ReplicatedEvalResult_AddSSTable, the Data
+	// field of which is an SST to be ingested (and which is present in memory
+	// but made durable via direct storage on the filesystem, bypassing the LSM).
 	RaftVersionSideloaded RaftCommandEncodingVersion = 1
+	// RaftVersionEmptyEntry is an empty entry. These are used by raft after
+	// leader election. Since they hold no data, there is nothing in them to
+	// decode.
+	RaftVersionEmptyEntry RaftCommandEncodingVersion = 253
+	// RaftVersionConfChange is a raftpb.Entry whose raftpb.EntryType is
+	// raftpb.EntryConfChange. The Entry's Data field holds a raftpb.ConfChange
+	// whose Context field is a kvserverpb.ConfChangeContext whose Payload is a
+	// kvserverpb.RaftCommand. In particular, the CmdIDKey requires a round of
+	// protobuf unmarshaling.
+	RaftVersionConfChange RaftCommandEncodingVersion = 254
+	// RaftVersionConfChangeV2 is analogous to RaftVersionConfChange, with
+	// the replacements raftpb.EntryConfChange{,V2} and raftpb.ConfChange{,V2}
+	// applied.
+	RaftVersionConfChangeV2 RaftCommandEncodingVersion = 255
 	// RaftCommandIDLen is the length for each command ID.
 	RaftCommandIDLen = 8
 	// RaftCommandPrefixLen is the prescribed length of each encoded command's prefix.
 	RaftCommandPrefixLen = 1 + RaftCommandIDLen
-	// RaftCommandNoSplitBit is deprecated.
-	// The no-split bit is now unused, but we still apply the mask to the first
-	// byte of the command for backward compatibility.
-	//
-	// TODO(tschottdorf): predates v1.0 by a significant margin. Remove.
-	RaftCommandNoSplitBit = 1 << 7
-	// RaftCommandNoSplitMask is deprecated.
-	RaftCommandNoSplitMask = RaftCommandNoSplitBit - 1
 )
 
 // EncodeRaftCommand encodes a raft command (including the versioning prefix).
```

Comment:
What types of command does this apply to? Same below.

Should be careful not to encode the new pseudo version values here?

Alternatively to overloading `RaftCommandEncodingVersion` (which now represents both the encoded version, and in-memory routing value), we could decouple the encoded consts and routing, i.e. have an in-memory only struct/int/byte (a pair of version and type), which is never serialized (but can be "converted" to serialized version by reading its version part or applying a mask).

Reply:
> Should be careful not to encode the new pseudo version values here?

I still want to formalize the encoding better, but not in this PR. Updated comment for now. These consts shouldn't even be exported.

> which now represents both the encoded version, and in-memory routing value)

We should talk about this because it came up in a few places. I would forget about versions. We can easily[^1] migrate between encodings using higher-level infrastructure that we have assuming we have a way to talk about encodings. So encodings are the thing to clarify, "versions" a thing to forget - the references to "versions" here date back to CRDB ca 2015 (the very beginnings!) and never made much sense but also there just wasn't any reason to touch it.

[^1]: not super easily, but easily enough, and what we do at the `kvserverbase` level doesn't really make a difference in how easy it is